src/algokit_utils/_debug_utils.py

```diff
@@ -0,0 +1,160 @@
+    program_hash = base64.b64encode(
+        checksum(base64.b64decode(result["result"]))  # type: ignore[no-untyped-call]
+    ).decode()
+    source_map = SourceMap(result["sourcemap"]).__dict__
+    source_map["sources"] = [file_name]
+
+    output_dir = output_path / ALGOKIT_DIR / DEBUGGER_DIR / app_name
+    source_map_output = output_dir / f'{file_name.replace(".teal", "")}.tok.map'
+    source_map_output.parent.mkdir(parents=True, exist_ok=True)
+    source_map_output.write_text(json.dumps(source_map))
+    (output_dir / file_name).write_text(teal_content)
+
+    return AVMDebuggerSourceMapEntry(str(source_map_output), program_hash)
+
+
+def persist_sourcemaps(
+    sources: list[PersistSourceMapInput],
+    project_root: Path,
+    client: "AlgodClient",
+) -> None:
+    sourcemaps = [
+        _build_avm_sourcemap(source.teal, source.app_name, source.file_name, project_root, client) for source in sources
+    ]
+
+    _upsert_debug_sourcemaps(sourcemaps, project_root)
+
+
+def simulate_response(atc: AtomicTransactionComposer, algod_client: "AlgodClient") -> SimulateAtomicTransactionResponse:
+    unsigned_txn_groups = atc.build_group()
+    empty_signer = EmptySigner()
+    txn_list = [txn_group.txn for txn_group in unsigned_txn_groups]
+    fake_signed_transactions = empty_signer.sign_transactions(txn_list, [])
+    txn_group = [SimulateRequestTransactionGroup(txns=fake_signed_transactions)]
+    trace_config = SimulateTraceConfig(enable=True, stack_change=True, scratch_change=True)
+
+    simulate_request = SimulateRequest(
+        txn_groups=txn_group, allow_more_logs=True, allow_empty_signatures=True, exec_trace_config=trace_config
+    )
+    return atc.simulate(algod_client, simulate_request)
+
+
+def simulate_and_persist_response(
+    atc: AtomicTransactionComposer, project_root: Path, algod_client: "AlgodClient"
+) -> None:
+    atc_to_simulate = atc.clone()
+    for txn_with_sign in atc_to_simulate.txn_list:
+        sp = algod_client.suggested_params()
+        txn_with_sign.txn.first_valid_round = sp.first
+        txn_with_sign.txn.last_valid_round = sp.last
+        txn_with_sign.txn.genesis_hash = sp.gh
+    response = simulate_response(atc_to_simulate, algod_client)
+    txn_types = [
+        txn_result["txn-results"][0]["txn-result"]["txn"]["txn"]["type"]
+        for txn_result in response.simulate_response["txn-groups"]
+    ]
+    txn_types_count = {txn_type: txn_types.count(txn_type) for txn_type in set(txn_types)}
+    txn_types_str = "_".join([f"{count}#{txn_type}" for txn_type, count in txn_types_count.items()])
+    last_round = response.simulate_response["last-round"]
+    output_file = (
+        project_root
+        / DEBUG_TRACES_DIR
+        / f'{datetime.now(tz=timezone.utc).strftime("%Y%m%d_%H%M%S")}_lr{last_round}_{txn_types_str}{TRACES_FILE_EXT}'
+    )
+    output_file.parent.mkdir(parents=True, exist_ok=True)
+    output_file.write_text(json.dumps(response.simulate_response, indent=2))
```

Comment:
Does this encode the Uint8Arrays into base64 strings or should we be using `get_obj_for_encoding`?

Reply:
yes, however we need to use the helper encoding method in typescript equivalent so ill have to be mindful when implementing it in ts utils